### PR TITLE
Develop unify

### LINKED ISSUE #34: Eliminate duplicate code between 3 vehicle "projects"
Currently the repo is made up of 3 different projects, one for each vehicle. The main.cpp and settings.h file in each of these three projects contains a lot of identical code. We will have to discuss how to optimize this, my thinking is to create another class in the library that contains all of the shared functionality in the main.cpp code.

The goal would be to have a single point where we could make changes that impact all 3 vehicles, instead of manually having to replicate our changes in each project.

### PR DESCRIPTION
**Pull request by silviu-toderita**
Unifies all three vehicle projects into one. Closes #34. The new folder structure is:

```
 lib/ - _external libraries_
 src/ - _telemetry source files_
    Logging/ - _All source files related to logging sensor data_
    Sensor/ - _All source files related to communicating with sensors_
    System/ - _All source files related to system tasks_
    main.cpp - _Common code for all 3 vehicles_
    settings.h - _Settings file (this is where you can choose the vehicle to compile and upload_
    vehicle.h - _Common vehicle header file_
    vehicleFc.cpp - _Code for fuel cell car_
    vehicleProto.cpp - _Code for proto car_
    vehicleUrban.cpp - _Code for urban car_ 
```

More comments in the code...

### CODE REVIEW

File: .github/workflows/build.yml
Comment (silviu-toderita) [inline .github/workflows/build.yml:5]: Continuous Integration modified to build only a single project

File: Makefile
Comment (silviu-toderita) [inline Makefile:6]: Continuous Integration modified to build only a single project

File: src/Logging/DataQueue.cpp
Comment (silviu-toderita) [inline src/Logging/DataQueue.cpp:27]: I was having some issues with including the settings.h file so I commented these out. I think they should be turned into getters instead, and we should do all debugging in the main or vehicle files.

File: src/Sensor/SensorGps.h
Comment (silviu-toderita) [inline src/Sensor/SensorGps.h:7]: Moved constants into GPS sensor header, as they're only needed here

File: src/System/Led.cpp
Comment (silviu-toderita) [inline src/System/Led.cpp:8]: Changed this enum as it was causing conflicts with constants that have since been removed. Better to have a more descriptive enum anyway.

File: src/System/Led.h
Comment (silviu-toderita) [inline src/System/Led.h:58]: Changed this enum as it was causing conflicts with constants that have since been removed. Better to have a more descriptive enum anyway.

File: src/build.mk
Comment (silviu-toderita) [inline src/build.mk:11]: Now all directories in /lib/ and /src/ are included, as well as all .cpp files in the first level of the directories in /lib/

Comment (silviu-toderita) [reply src/build.mk:11]: This is what enables us to move all of our code into source, by creating directories to organize the different parts of our project.

File: src/main.cpp
Comment (silviu-toderita) [inline src/main.cpp:27]: We haven't been making use of this functionality and it's become less useful over time because we're doing less in the "handle" functions of each sensor. I've left the underlying functions alone, but removed it from the main file for now until we find a better way to run benchmarks

File: src/vehicle.h
Comment (silviu-toderita) [inline src/vehicle.h:1]: Lots of duplicate functionality in the vehicle headers, so it made sense to combine them into one.

File: src/vehicleFc.cpp
Comment (silviu-toderita) [inline src/vehicleFc.cpp:3]: Each vehicle's cpp file is only compiled if that vehicle is selected.

File: src/vehicleProto.cpp
Comment (silviu-toderita) [inline src/vehicleProto.cpp:3]: Each vehicle's cpp file is only compiled if that vehicle is selected.

File: src/vehicleUrban.cpp
Comment (silviu-toderita) [inline src/vehicleUrban.cpp:3]: Each vehicle's cpp file is only compiled if that vehicle is selected.

File: src/main.cpp
Comment (silviu-toderita) [inline src/main.cpp:44]: Check to make sure a correct vehicle has been chosen. If not, project should not compile.

File: .github/workflows/build.yml
Comment (lcfyi) [reply .github/workflows/build.yml:5]: Is there any reason why we aren't building for all vehicles? It looks like the default builds for proto, but if we make changes to another vehicle, we'd lose the static analysis from those builds.

File: src/Logging/DataQueue.cpp
Comment (lcfyi) [reply src/Logging/DataQueue.cpp:27]: What was the issue you were running into? I tried a minimal build with this line commented out and `settings.h` included and didn't come across any issues.

File: src/System/Led.cpp
Comment (lcfyi) [inline src/System/Led.cpp:42]: This can be rewritten as a switch statement; would make the code a bit cleaner.

File: src/main.cpp
Comment (lcfyi) [reply src/main.cpp:27]: We could probably put the publish methods with the data queue since they're closely related anyway.

File: src/main.cpp
Comment (lcfyi) [reply src/main.cpp:44]: I don't think we need this static assert since we have the define guards in the implementations; if none of the implementations are selected, the linker won't be able to find proper definitions and will fail to compile.

File: src/settings.h
Comment (lcfyi) [inline src/settings.h:49]: Instead of using another variable to select the current vehicle, what do you think of just defining macros for each vehicle? That way we can pass a compiler directive like `-DPROTO`, which our code can then surround with:

```cpp
#ifdef PROTO
// proto stuff
#endif
```

File: src/vehicle.h
Comment (lcfyi) [inline src/vehicle.h:17]: If we're going to combine the vehicle headers into one, we should also remove these and put them in the implementations themselves.

File: src/vehicleFc.cpp
Comment (lcfyi) [reply src/vehicleFc.cpp:3]: I added a comment in the settings file, but we can simplify this with an `ifdef` for the particular vehicle.

File: src/vehicleFc.cpp
Comment (lcfyi) [inline src/vehicleFc.cpp:24]: Can we put these as macros or in the array type itself?

**Review by lcfyi — COMMENTED** [review]
I left some comments.

I do like the reduced verbosity/duplication by merging a lot of the code, but it does feel a bit clumsier to have to change a macro every time you want to compile and build for another vehicle with the workbench. However, if you're not using that then it's not a huge deal to pass directives in the compilation command; if you guys don't think that's a downside to your development workflow then that's totally fine.

File: .github/workflows/build.yml
Comment (silviu-toderita) [reply .github/workflows/build.yml:5]: Great point! Is there a way to pass in a constant to be able to compile 3 different sources?

File: src/main.cpp
Comment (silviu-toderita) [reply src/main.cpp:27]: True, I think I will leave it to another issue though as we were discussing publishing when the buffer gets (close to) full as opposed to on a set interval.

File: src/Logging/DataQueue.cpp
Comment (silviu-toderita) [reply src/Logging/DataQueue.cpp:27]: It looks like it's now compiling fine, I think I was in makefile hell yesterday and it was a separate issue altogether. I'll revert this.

File: src/System/Led.cpp
Comment (silviu-toderita) [reply src/System/Led.cpp:42]: Fixed!

File: src/main.cpp
Comment (silviu-toderita) [reply src/main.cpp:44]: I've removed it

File: .github/workflows/build.yml
Comment (lcfyi) [reply .github/workflows/build.yml:5]: You can with compiler flags. The DeviceOS makefile accepts an argument `EXTRA_CFLAGS` which you can use to pass in macros to the compiler.

For example, if you wanted to pass in `CURRENT_VEHICLE` as 3 and to define `DEBUG`, you can do:

```bash
make all PLATFORM=boron APPDIR=/app/ TARGET_DIR=/$(OUTPUT_DIR) EXTRA_CFLAGS='-DCURRENT_VEHICLE=3 -DDEBUG=0'
```

Here's a [good reference](https://github.com/particle-iot/device-os/blob/develop/docs/build.md#variables).

File: src/vehicle.h
Comment (silviu-toderita) [reply src/vehicle.h:17]: gps is still required in the main file as it also does timekeeping duties, so I think we should either extern it here or create the object in main.cpp

File: .github/workflows/build.yml
Comment (silviu-toderita) [reply .github/workflows/build.yml]: Done! I've made it so you can pass a CFLAG in for the vehicle or define it in the settings file. Both work, and I prefer the latter. CI now compiles all 3 vehicles again

File: src/vehicleFc.cpp
Comment (silviu-toderita) [reply src/vehicleFc.cpp]: Done!

File: src/vehicleFc.cpp
Comment (silviu-toderita) [reply src/vehicleFc.cpp:24]: I don't see a way to do that without putting the sensor count in a different file. Do you have any suggestions for how to do this cleanly?

File: Makefile
Comment (silviu-toderita) [reply Makefile]: Now changed to compile all 3 vehicles again

**Comment by silviu-toderita** [comment]
@cosparks Any comments on this PR before I merge it? I've made a bunch of changes to stuff that you just did, so want to make sure you're good with it.

File: src/vehicleFc.cpp
Comment (silviu-toderita) [reply src/vehicleFc.cpp:24]: Created issue #61 for this

File: src/vehicle.h
Comment (cosparks) [reply src/vehicle.h:17]: we need gps and dataqueue in both main.cpp and globals.cpp.  can't see a way of doing that without including a .cpp file in one or the other, or declaring them as extern

File: .github/workflows/build.yml
Comment (lcfyi) [inline .github/workflows/build.yml:18]: Could probably remove this conditional since it's not very accurate at detecting vehicle-specific changes now.

It's not a huge deal since builds run in parallel and the minutes are free for public repos.

**Comment by cosparks** [comment]
Did you make any changes to my work besides moving things around and changing #includes stuff?

It all looks good to me from what I've seen!

File: src/vehicle.h
Comment (lcfyi) [reply src/vehicle.h:17]: > gps is still required in the main file as it also does timekeeping duties, so I think we should either extern it here or create the object in main.cpp

Can we make this a virtual function in the `vehicle.h` header to retrieve the time, which the vehicles can implement based on the existence of a GPS? I think it would be better to not need to rely on specific sensors, as it would make the code a lot more portable as well.



> we need gps and dataqueue in both main.cpp and globals.cpp. can't see a way of doing that without including a .cpp file in one or the other, or declaring them as extern

I think it would be better if we don't put them in main since I would consider their inclusions something that may have side effects. Since we're merging the vehicle-specific code to a single main file, we'd probably want to abstract as much as we can from the entrypoint so it's easier to isolate changes from one vehicle's code to another.

File: src/vehicleFc.cpp
Comment (lcfyi) [reply src/vehicleFc.cpp:24]: > I don't see a way to do that without putting the sensor count in a different file. Do you have any suggestions for how to do this cleanly?

There was some residual discussion about this in #50, but I think using C++ STL containers would be the way to go (so either an `array` or `vector`).

File: .github/workflows/build.yml
Comment (silviu-toderita) [reply .github/workflows/build.yml]: thanks, done!

File: src/vehicle.h
Comment (silviu-toderita) [reply src/vehicle.h:17]: Good idea, I'll add this to issue #48

File: src/main.cpp
Comment (lcfyi) [inline src/main.cpp:82]: Looks like we had a casualty in the latest commit 😛

**Review by lcfyi — APPROVED** [review]

File: src/main.cpp
Comment (silviu-toderita) [reply src/main.cpp:85]: Haha oops!

**Pull request merged**
